qiskit/circuit/library/arithmetic/multipliers/hrs_cumulative_multiplier.py

```diff
@@ -0,0 +1,134 @@
+
+    **References:**
+
+    [1] Häner et al., Optimizing Quantum Circuits for Arithmetic, 2018.
+    `arXiv:1805.12445 <https://arxiv.org/pdf/1805.12445.pdf>`_
+
+    """
+
+    def __init__(
+        self,
+        num_state_qubits: int,
+        num_result_qubits: Optional[int] = None,
+        adder: Optional[QuantumCircuit] = None,
+        name: str = "HRSCumulativeMultiplier",
+    ) -> None:
+        r"""
+        Args:
+            num_state_qubits: The number of qubits in either input register for
+                state :math:`|a\rangle` or :math:`|b\rangle`. The two input
+                registers must have the same number of qubits.
+            num_result_qubits: The number of result qubits to limit the output to.
+                If number of result qubits is :math:`n`, multiplication modulo :math:`2^n` is performed
+                to limit the output to the specified number of qubits. Default
+                value is ``2 * num_state_qubits`` to represent any possible
+                result from the multiplication of the two inputs.
+            adder: Half adder circuit to be used for performing multiplication. The
+                CDKMRippleCarryAdder is used as default if no adder is provided.
+            name: The name of the circuit object.
+        Raises:
+            NotImplementedError: If ``num_result_qubits`` is not default and a custom adder is provided.
+        """
+        super().__init__(num_state_qubits, num_result_qubits, name=name)
+
+        if self.num_result_qubits != 2 * num_state_qubits and adder is not None:
+            raise NotImplementedError("Only default adder is supported for modular multiplication.")
+
+        # define the registers
+        qr_a = QuantumRegister(num_state_qubits, name="a")
+        qr_b = QuantumRegister(num_state_qubits, name="b")
+        qr_out = QuantumRegister(self.num_result_qubits, name="out")
+        self.add_register(qr_a, qr_b, qr_out)
+
+        # prepare adder as controlled gate
+        if adder is None:
+            from qiskit.circuit.library.arithmetic.adders import CDKMRippleCarryAdder
+
+            adder = CDKMRippleCarryAdder(num_state_qubits, kind="half")
+
+        # get the number of helper qubits needed
+        num_helper_qubits = adder.num_ancillas
+
+        # add helper qubits if required
+        if num_helper_qubits > 0:
+            qr_h = AncillaRegister(num_helper_qubits, name="helper")  # helper/ancilla qubits
+            self.add_register(qr_h)
+
+        # build multiplication circuit
+        for i in range(num_state_qubits):
+            excess_qubits = max(0, num_state_qubits + i + 1 - self.num_result_qubits)
+            if excess_qubits == 0:
+                num_adder_qubits = num_state_qubits
+                adder_for_current_step = adder
+            else:
+                num_adder_qubits = num_state_qubits - excess_qubits + 1
+                adder_for_current_step = CDKMRippleCarryAdder(num_adder_qubits, kind="fixed")
```

Comment:
This part will only be reached if the number of result qubits is not 2n, right?

Reply:
Yes, since `excess_qubits` can have non-zero values only when `num_result_qubits < 2 * num_state_qubits` 🙂